File: python/mirage/commands/builtin/generic/readlink.py
import posixpath

from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


async def readlink(
    paths: list[PathSpec],
    *,
    f: bool = False,
    e: bool = False,
    m: bool = False,
    n: bool = False,
) -> tuple[ByteSource | None, IOResult]:
    if not paths:
        raise ValueError("readlink: missing operand")
    normalize = f or e or m
    results: list[str] = []
    for p in paths:
        vp = p.prefix + "/" + p.original.lstrip(
            "/") if p.prefix else p.original
        if normalize:
            vp = posixpath.normpath(vp)
        results.append(vp)
    text = "\n".join(results)
    if not n:
        text += "\n"
    return text.encode(), IOResult()


__all__ = ["readlink"]

File: typescript/packages/core/src/commands/builtin/generic/readlink.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { IOResult, type ByteSource } from '../../../io/types.ts'
import type { PathSpec } from '../../../types.ts'
import type { CommandFnResult, CommandOpts } from '../../config.ts'

const ENC = new TextEncoder()

function normPath(p: string): string {
  const parts = p.split('/')
  const out: string[] = []
  for (const seg of parts) {
    if (seg === '' || seg === '.') continue
    if (seg === '..') {
      if (out.length > 0) out.pop()
      continue
    }
    out.push(seg)
  }
  const leading = p.startsWith('/') ? '/' : ''
  return leading + out.join('/') || (leading !== '' ? '/' : '.')
}

export function readlinkGeneric(
  paths: PathSpec[],
  _texts: string[],
  opts: CommandOpts,
): CommandFnResult {
  if (paths.length === 0) {
    return [null, new IOResult({ exitCode: 1, stderr: ENC.encode('readlink: missing operand\n') })]
  }
  const normalize = opts.flags.f === true || opts.flags.e === true || opts.flags.m === true
  const noNewline = opts.flags.n === true
  const results: string[] = []
  for (const p of paths) {
    let vp = p.prefix !== '' ? p.prefix + '/' + p.original.replace(/^\/+/, '') : p.original
    if (normalize) vp = normPath(vp)
    results.push(vp)
  }
  let text = results.join('\n')
  if (!noNewline) text += '\n'
  const out: ByteSource = ENC.encode(text)
  return [out, new IOResult()]
}
